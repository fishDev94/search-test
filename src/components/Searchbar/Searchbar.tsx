import { ChangeEvent, useCallback, useMemo, useState, useEffect } from "react";
import AppButton from "../UI/AppButton/AppButton";
import styles from "./searchbar.module.scss";
import useSearchContacts from "../../hook/useSearchContacts";
import { Comment } from "../../types";
import SearchResults from "../SearchResults/SearchResults";

export default function Searchbar({
  value,
  handleSearch = () => {},
  onSumbit = () => {},
  onClear = () => {},
}: {
  value: string;
  handleSearch: (arg: string) => void;
  onSumbit: () => void;
  onClear: () => void;
}) {
  const [isSearching, setIsSearching] = useState(false);
  const [isSearchResultsOpened, setIsSearchResultsOpened] = useState(false);

  useEffect(() => {
    setIsSearching(value.length >= 3);
    setIsSearchResultsOpened(value.length >= 3);
  }, [value]);

  const handleSearchInput = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    handleSearch(value);
  };

  const { data } = useSearchContacts<Comment[]>(value, {
    enabled: isSearching,
    isSubmitted: isSearching,
    key: "searchbar",
  });

  const onSumbitClick = () => {
    onSumbit();
    setIsSearchResultsOpened(false);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter" && isSearching) {
      onSumbitClick();
    }
  };

  const isSearchResults = useMemo(() => {
    return data ? data.length > 0 : false;
  }, [data]);

  const clearValue = () => {
    handleSearch("");
    setIsSearching(false);
    setIsSearchResultsOpened(false);
    onClear();
  };

  const clearButton = useCallback(() => {
    return (
      value.length > 0 && (
        <span
          onClick={clearValue}
          className={styles.searchbar__container_clear}
          data-testid="clear-button"
        >
          clear
        </span>
      )
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);

  return (
    <div className={styles.searchbar}>
      <div className={styles.searchbar__container}>
        <input
          className={styles.searchbar__container_input}
          onInput={handleSearchInput}
          type="text"
          placeholder="search"
          onKeyDown={handleKeyDown}
          value={value}
        />
        {clearButton()}
        <AppButton
          text="Search"
          onClick={onSumbitClick}
          disabled={!isSearching}
        />
      </div>
      <SearchResults
        isActive={isSearchResults && isSearchResultsOpened}
        searchResultData={data}
      />
    </div>
  );
}
