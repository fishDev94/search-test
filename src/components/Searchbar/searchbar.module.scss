.searchbar {
    display: flex;
    gap: 12px;
    box-shadow: var(--box-shadow-4dp);
    border-radius: var(--border-radius-m);
    padding: 4px 8px;
    overflow: hidden;
    width: max-content;

    &__input {
        border: 0;
        outline: 0;
    }
}