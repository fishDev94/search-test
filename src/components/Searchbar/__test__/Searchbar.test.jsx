import { describe, it, expect, vi } from "vitest";
import { render, fireEvent } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import Searchbar from "../Searchbar";

vi.mock("../../hook/useSearchContacts", () => ({
  default: vi.fn(() => ({ data: [] })),
}));

vi.mock("../SearchResults/SearchResults", () => ({
  default: () => <div data-testid="search-results" />,
}));

const queryClient = new QueryClient();
const wrapper = ({ children }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

describe("Searchbar", () => {
  const mockProps = {
    value: "",
    handleSearch: vi.fn(),
    onSumbit: vi.fn(),
    onClear: vi.fn(),
  };

  it("matches snapshot", () => {
    const { container } = render(<Searchbar {...mockProps} />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  it("updates input value", () => {
    const { container } = render(<Searchbar {...mockProps} />, { wrapper });
    const input = container.querySelector("input");

    fireEvent.input(input, { target: { value: "test" } });
    expect(mockProps.handleSearch).toHaveBeenCalledWith("test");
  });

  it("shows clear button with value", () => {
    const { getByTestId } = render(<Searchbar {...mockProps} value="test" />, {
      wrapper,
    });

    const clearButton = getByTestId("clear-button");
    expect(clearButton).toBeTruthy();
    expect(clearButton.textContent).toBe("clear");
  });

  it("enables button after 3 chars", () => {
    const { container, rerender } = render(
      <Searchbar {...mockProps} value="12" />,
      { wrapper }
    );
    const button = container.querySelector("button");
    expect(button.disabled).toBe(true);

    rerender(<Searchbar {...mockProps} value="123" />, { wrapper });
    expect(button.disabled).toBe(false);
  });
});
