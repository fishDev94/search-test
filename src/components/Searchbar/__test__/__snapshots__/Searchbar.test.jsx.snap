// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Searchbar > matches snapshot 1`] = `
<div>
  <div
    class="_searchbar_cc4f57"
  >
    <div
      class="_searchbar__container_cc4f57"
    >
      <input
        class="_searchbar__container_input_cc4f57"
        placeholder="search"
        type="text"
        value=""
      />
      <button
        class="_app_button_1f8b9b _inactive_1f8b9b"
        disabled=""
        name="app_button"
      >
        Search
      </button>
    </div>
    <ul
      class="_search_results_99e224 "
    />
  </div>
</div>
`;
